[InstCombine] Add limit for expansion of gep chains

nikic opened the pull request:
When converting gep subtraction / comparison to offset subtraction / comparison, avoid expanding very long multi-use gep chains.

nikic committed:
[InstCombine] Add limit for expansion of gep chains

When converting gep subtraction / comparison to offset
subtraction / comparison, avoid expanding very long multi-use
gep chains.

Another improvement we could make is to expand one-use followed
by multi-use gep differently, by rewriting the multi-use gep to
include the one-use offsets. But I think we want to have some
kind of complexity cut-off in any case.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:
```diff
@@ -2144,13 +2144,35 @@ CommonPointerBase CommonPointerBase::compute(Value *LHS, Value *RHS) {
+    bool SeenMultiUse = false;
+    for (GEPOperator *GEP : GEPs) {
+      // Only count GEPs after the first multi-use GEP. For the first one,
+      // we will directly reuse the offset.
```
> Do you think it would be better to start with changing the GEP expansion/rewriting?

Yeah, if the compile-time impact is acceptable.

nikic committed:
Update for new one-use GEP expansion

nikic committed:
Adjust tests for new behavior

nikic committed:
Remove constant special case

With the new expansion this would have to be more careful about
a constant on multi-use GEP and non-constant on merged one-use
GEPs. Just drop this heuristic for now.

nikic commented:
After github.com/llvm/llvm-project/pull/147263 this doesn't have all that much effect anymore, but it should at least avoid any pathological cases.

dtcxzyw reviewed (approved):
LG
